src/components/forms/FileInput/FileInput.tsx

```diff
@@ -12,164 +12,186 @@ interface FileInputProps {
+    )
+
+    const targetClasses = classnames('usa-file-input__target', {
+      'usa-file-input--drag': isDragging,
+      'has-invalid-file': showError,
+    })
+
+    const hideDragText =
+      /rv:11.0/i.test(navigator.userAgent) ||
+      /Edge\/\d./i.test(navigator.userAgent)
+
+    const dragText = multiple ? 'Drag files here or ' : 'Drag file here or '
+
+    const filePreviews = []
+    if (files) {
+      for (let i = 0; i < files?.length; i++) {
+        const imageId = makeSafeForID(files[parseInt(`${i}`)].name)
+        const key = `filePreview_${imageId}`
+        filePreviews.push(
+          <FilePreview
+            key={key}
+            imageId={imageId}
+            file={files[parseInt(`${i}`)]}
+          />
+        )
       }
+    }
 
-      if (!allFilesAllowed) {
-        setFiles(null)
-        setShowError(true)
-        e.preventDefault()
-        e.stopPropagation()
+    const instructionClasses = classnames('usa-file-input__instructions', {
+      'display-none': filePreviews.length > 0,
+    })
+
+    const previewHeaderText =
+      filePreviews.length > 1
+        ? `${filePreviews.length} files selected`
+        : 'Selected file'
+
+    const preventInvalidFiles = (e: React.DragEvent): void => {
+      setShowError(false)
+
+      if (accept) {
+        const acceptedTypes = accept.split(',')
+        let allFilesAllowed = true
+        for (let i = 0; i < e.dataTransfer.files.length; i += 1) {
+          const file = e.dataTransfer.files[parseInt(`${i}`)]
+          if (allFilesAllowed) {
+            for (let j = 0; j < acceptedTypes.length; j += 1) {
+              const fileType = acceptedTypes[parseInt(`${j}`)]
+              allFilesAllowed =
+                file.name.indexOf(fileType) > 0 ||
+                file.type.includes(fileType.replace(/\*/g, ''))
+              if (allFilesAllowed) break
+            }
+          } else break
+        }
+
+        if (!allFilesAllowed) {
+          setFiles([])
+          setShowError(true)
+          e.preventDefault()
+          e.stopPropagation()
+        }
       }
     }
-  }
 
-  // Event handlers
-  const handleDragOver = (): void => setIsDragging(true)
-  const handleDragLeave = (): void => setIsDragging(false)
-  const handleDrop = (e: React.DragEvent): void => {
-    preventInvalidFiles(e)
-    setIsDragging(false)
-    if (onDrop) onDrop(e)
-  }
+    // Event handlers
+    const handleDragOver = (): void => setIsDragging(true)
+    const handleDragLeave = (): void => setIsDragging(false)
+    const handleDrop = (e: React.DragEvent): void => {
+      preventInvalidFiles(e)
+      setIsDragging(false)
+      if (onDrop) onDrop(e)
+    }
 
-  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
-    setShowError(false)
-    setFiles(e.target?.files)
-    if (onChange) onChange(e)
-  }
+    const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+      setShowError(false)
+      setFiles(e.target?.files || [])
```

Comment:
This is causing issues in my code because I handle only FileList which is the behavior of the normal input.

Reply:
Also, there may be state update issue. On initial drop, `fileInputRef.current` consistently doesn't register the files. Only on the second drop, when a `FileList` is returned,  does it register that the files have changed.  

I'll push up my code so you can see it in our storybook (which builds by branch).

Reply:
the state weirdness was resolved by only referencing ` fileInputRef.current.input` instead.